### DISCUSSION TITLE
How can I query for properties of objects in a Revit file?

### DISCUSSION THREAD

**Discussion by barhamd** [iTwin/itwinjs-core#94] [General] [answered]
I'm trying to get a list of properties for each object in a Revit file. Something similar to this (from [Forge's Derivatives API](https://forge.autodesk.com/en/docs/model-derivative/v2/reference/http/urn-metadata-guid-properties-GET/)):

```JSON
   {
        "objectid": 1,
        "name": "A5",
        "externalId": "mou0zG8ViUOsqUzhb4TUiA",
        "properties": {
          "Name": "A5"
        }
      },
      {
        "objectid": 2,
        "name": "Model",
        "externalId": "z4u0zG8ViUOsqUzhb4TUiA",
        "properties": {
          "Component Name": "Model",
          "Name": "Model",
          "Design Tracking Properties": {
            "Design State": "WorkInProgress",
            "Designer": "ADSK",
            "File Subtype": "Assembly"
          },
          "File Properties": {
            "Author": "ADSK",
            "Creation Date": "2012-Jul-09 20:18:20",
            "Original System": "Autodesk Inventor 2017",
            "Part Number": "Model"
          },
          "Mass Properties": {
            "Area": "19772.676 millimeter^2",
            "Volume": "83673.946 millimeter^3"
          }
        }
      },
```

I've attempted to use the [iModelConsole](https://imodelconsole.bentley.com/) to run some ECSQL queries but haven't been able to find the right query.

I see that there is an example in the [Simple Viewer App](https://github.com/imodeljs/imodeljs-samples/tree/master/interactive-app/simple-viewer-app), however I've been unable to explore this. The Revit files I've tried successfully load, but I receive an `Unhandled Rejection (InvalidArgument): Invalid content field category` error when I try to select an object from the tree. It also appears to be a "Presentation" component, so I'm not sure if it's the wrong path.

How can I accomplish this?

**Comment by jan-slegr** [comment]
> I'm trying to get

What method do you prefer? Is your question how to obtain data using ECSQL (e.g. iModelConsole) or using iModelJS API? I think any ECSQL query should be tested in iModelConsole at first, before it's implemented in a code.

> a list of properties

I know it's a bit of meticulous comment, but I guess you don't want _a list of properties_, but _a list of properties values_. ;-) The list of properties is defined by a class structure, whereas the properties values represent data stored with the object.

> Something similar to this (from [Forge's Derivatives API](https://forge.autodesk.com/en/docs/model-derivative/v2/reference/http/urn-metadata-guid-properties-GET/)):

I think it's hard to compare, because as far as I understand, Forge API is REST API, whereas iModelJS is TypeScript library, that can be used to communicate with iModelHub. I think iModelHub does not publish any public REST API that can be used to obtain model data directly.

> How can I accomplish this?

I have not any imodel created from Revit data available, so I do not know its structure and how conversion was done (and what schemas were used), so my recommendation is general. It's based on assumption you have the imodel available at iModelHub.

1. Use iModel.js Design Review to check the model visually. An advantage of this tool is you can select an element and available information are displayed, so it's easy to obtain information using what classes the element is described.
1. Open the model in iModelConsole.
1. From the console, open [iModel Schema Explorer](https://imodelschemaeditor.bentley.com/). When a model is opened in the console, you can open the explorer using icon at top left corner.
1. "A mental way" to construct right ECSQL is in my opinion: select element in Design Review to check used classes > in Schema Explorer discover the class(es) structure > create ECSQL.

I recommend to see [ECSQL tutorial](https://www.imodeljs.org/learning/ecsqltutorial/) to find typical queries, but I assume it will be similar to a standard basic query structure:

`select * from <alis>.<class>`

To select all generic elements, when a plain geometry was converted to iModel and they are not described using special classes:

`select * from Generic.PhysicalObject`

With regards,

  Jan

**Comment by grigasp** [comment, marked as answer]
Using ECSQL for getting all properties of an element is really not an option, because majority of the properties are usually located on related [element aspects](https://www.imodeljs.org/bis/intro/elementaspect-fundamentals/), [types](https://www.imodeljs.org/bis/intro/type-definitions/), etc.. To simplify getting iModel data in a user and developer -friendly way we suggest using the [Presentation API](https://www.imodeljs.org/learning/presentation/) which looks up for all the related properties for you. 

You can find an example of how to use the API directly [here](https://gist.github.com/grigasp/aa75511dc4137137b651f56676c27f87) - have a look at the `listElementProperties` method. Note that the example is for backend code, but the API on the frontend is nearly identical.

I also checked the simple-viewer-app and was able to reproduce the issue you're seeing. It's caused by the `Table` component in the app - if you comment-out the [line 297 in App.tsx](https://github.com/imodeljs/imodeljs-samples/blob/81af4cd3d3344caa124c50e7ddfac99158bc26a9/interactive-app/simple-viewer-app/src/frontend/components/App.tsx#L297) the app should start working fine and you'll see properties in the properties widget. We'll work on a fix for the Table component ASAP.
